### PR TITLE
Feature/pull endpoint

### PR DESCRIPTION
**Pull request by gunther-bachmann**

### CODE REVIEW

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend [hunk 1]
```diff
@@ -36,29 +35,90 @@ class DocumentProvider {
...
+			throw new ExistingFileException('''target file '«newPath»' does already exist''')
+		} else {
+			val preCommit = git.repository.resolve('HEAD')
+			logger.trace('commitid before any action is taken is ' + preCommit.getName)
+			try {
+				copySuccessful = doCleanCopy(file, newFile, pushAction)
+			} catch (Exception e) {
+				logger.error('exception during copy action', e)
+			}
+			if (!copySuccessful) {
```
Comment (jjelschen) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend:70]: `doCleanCopy(…)` never returns `false` – its either successful, or it throws an exception

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend [hunk 2]
```diff
@@ -283,27 +315,9 @@ class DocumentProvider {
...
-			throw new MaliciousPathException(workspacePath, filePath, userProvider.get.name)
-		}
-	}
-
-	private def boolean create(File file) {
-		val parent = new File(file.parent)
-		if (!parent.exists) {
-			logger.debug("Creating directory='{}'.", parent)
-			parent.mkdirs
+			logger.
```
Comment (jjelschen) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend]: formatter fail?

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/workspace/WorkspaceProvider.xtend [hunk 3]
```diff
@@ -1,16 +1,34 @@
...
 import java.io.File
 import javax.inject.Inject
 import javax.inject.Provider
+import org.slf4j.LoggerFactory
 import org.testeditor.web.backend.persistence.PersistenceConfiguration
+import org.testeditor.web.backend.persistence.exception.MaliciousPathException
 import org.testeditor.web.dropwizard.auth.User
 
+import static java.nio.charset.StandardCharsets.UTF_8
+import com.google.common.annotations.VisibleForTesting
```
Comment (jjelschen) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/workspace/WorkspaceProvider.xtend]: Organize imports! :satisfied:

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/workspace/WorkspaceResource.xtend [hunk 4]
```diff
@@ -25,6 +41,110 @@ class WorkspaceResource {
...
+	 * 2. create a backup file for dirtyResources that have changed through the pull
+	 * 
+	 * Note that the created backup files do not hold the contents of the dirtyResources,
+	 * since their contents are known to the front end only! The backup files created will hold the
+	 * same content as the respective file just pulled (as a default).
+	 * 
+	 * The frontend is expected to react accordingly, that is:
+	 * 1. it informs the user about resources that have changed
+	 * 2. it provides some resolution/information for files the user has changed and that have changed
+	 *    in the repository, too. The user must be given the chance to persisted his local changes into
```
Comment (jjelschen) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/workspace/WorkspaceResource.xtend]:
```suggestion
	 *    in the repository, too. The user must be given the chance to persist his local changes into 
```

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/workspace/WorkspaceResource.xtend [hunk 5]
```diff
@@ -25,6 +41,110 @@ class WorkspaceResource {
...
+				it.backupResource = backupFile
+			])
+		}
+	}
+
+	/** create a tree iterator (for diff calculation) on given objectId (commit) */
+	private static def AbstractTreeIterator prepareTreeParser(Repository repository,
+		ObjectId objectId) throws IOException {
+		// from the commit we can build the tree which allows us to construct the TreeParser
+		// noinspection Duplicates
```
Comment (jjelschen) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/workspace/WorkspaceResource.xtend]: is this (still) needed?

**Review by jjelschen — CHANGES_REQUESTED** [review]

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend (hunk 1 above)
Comment (gunther-bachmann) [reply org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/DocumentProvider.xtend:70]: corrected that, kept the variable anyway, made the following code easier

**Review by jjelschen — APPROVED** [review]

**Pull request merged**